src/azure-cli-core/azure/cli/core/telemetry.py

```diff
@@ -0,0 +1,133 @@
+#---------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+#---------------------------------------------------------------------------------------------
+from __future__ import print_function
+import getpass
+import time
+import subprocess
+import json
+import os
+import sys
+
+from applicationinsights import TelemetryClient
+from applicationinsights.exceptions import enable
+from azure.cli.core import __version__ as core_version
+
+_DEBUG_TELEMETRY = 'AZURE_CLI_DEBUG_TELEMETRY'
+client = {}
+data = []
+APPLICATION = {}
+ARGCOMPLETE_ENV_NAME = None
+
+def set_application(application, argcomplete_env_name):
+    global APPLICATION #pylint: disable=global-statement
+    APPLICATION = application
+    global ARGCOMPLETE_ENV_NAME #pylint: disable=global-statement
+    ARGCOMPLETE_ENV_NAME = argcomplete_env_name
+
+def init_telemetry():
+    try:
+        instrumentation_key = '02b91c82-6729-4241-befc-e6d02ca4fbba'
+
+        global client #pylint: disable=global-statement
+        client = TelemetryClient(instrumentation_key)
+
+        client.context.application.id = 'Azure CLI'
+        client.context.application.ver = core_version
+        client.context.user.id = hash(getpass.getuser())
+
+        enable(instrumentation_key)
+    except Exception as ex: #pylint: disable=broad-except
+        # Never fail the command because of telemetry, unless debugging
+        if _debugging():
+            raise ex
+
+def user_agrees_to_telemetry():
+    # TODO: agreement, needs to take Y/N from the command line
+    # and needs a "skip" param to not show (for scripts)
+    return True
+
+def log_event(event_name, **kwargs):
+    """
+    IMPORTANT: do not log events with quotes in the name, properties or measurements;
+    those events may fail to upload.  Also, telemetry events must be verified in
+    the backend because successful upload does not guarentee success.
+    """
+    try:
+        event_name = _remove_quotes(event_name)
+        for key, value in kwargs.items():
+            kwargs[key] = _remove_quotes(value)
+
+        source = 'az'
+        if os.environ.get('CONTINUOUS_INTEGRATION') and os.environ.get('TRAVIS'):
+            source = 'CI'
+        elif ARGCOMPLETE_ENV_NAME in os.environ:
+            source = 'completer'
```

Comment:
the OS info is auto-captured by app insights, although we could fill this out with other values if needed

---
In reply to: [78660718](https://github.com/Azure/azure-cli/pull/865#discussion_r78660718) [](ancestors = 78660718)